// ylru/dist/commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LRU = void 0;
class LRU {
    #max;
    #size = 0;
    #cache = new Map();
    #_cache = new Map();
    constructor(max) {
        this.#max = max;
    }
    /** only for unittest */
    get cache() {
        return this.#cache;
    }
    /** only for unittest */
    get _cache() {
        return this.#_cache;
    }
    get(key, options) {
        let item = this.#cache.get(key);
        const maxAge = options?.maxAge;
        // only call Date.now() when necessary
        let now;
        function getNow() {
            now = now || Date.now();
            return now;
        }
        if (item) {
            // check expired
            if (item.expired && getNow() > item.expired) {
                item.expired = 0;
                item.value = undefined;
            }
            else {
                // update expired in get
                if (maxAge !== undefined) {
                    const expired = maxAge ? getNow() + maxAge : 0;
                    item.expired = expired;
                }
            }
            return item.value;
        }
        // try to read from _cache
        item = this.#_cache.get(key);
        if (item) {
            // check expired
            if (item.expired && getNow() > item.expired) {
                item.expired = 0;
                item.value = undefined;
            }
            else {
                // not expired, save to cache
                this._update(key, item);
                // update expired in get
                if (maxAge !== undefined) {
                    const expired = maxAge ? getNow() + maxAge : 0;
                    item.expired = expired;
                }
            }
            return item.value;
        }
    }
    set(key, value, options) {
        const maxAge = options?.maxAge;
        const expired = maxAge ? Date.now() + maxAge : 0;
        let item = this.#cache.get(key);
        if (item) {
            item.expired = expired;
            item.value = value;
        }
        else {
            item = {
                value,
                expired,
            };
            this._update(key, item);
        }
    }
    keys() {
        const cacheKeys = new Set();
        const now = Date.now();
        for (const [key, item] of this.#cache.entries()) {
            checkEntry(key, item);
        }
        for (const [key, item] of this.#_cache.entries()) {
            checkEntry(key, item);
        }
        function checkEntry(key, item) {
            if (item.value && (!item.expired) || item.expired >= now) {
                cacheKeys.add(key);
            }
        }
        return Array.from(cacheKeys);
    }
    reset() {
        this.#size = 0;
        this.#cache.clear();
        this.#_cache.clear();
    }
    _update(key, item) {
        this.#cache.set(key, item);
        this.#size++;
        if (this.#size >= this.#max) {
            this.#size = 0;
            this.#_cache = this.#cache;
            this.#cache = new Map();
        }
    }
}
exports.LRU = LRU;
