// bitdepth/dist/bitdepth.umd.js
(function (global, factory) {typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :typeof define === 'function' && define.amd ? define(factory) :(global.bitDepth = factory());}(this, (function () { 'use strict';var bitDepth=function(){/** @private @const */ var f64f32_=new Float32Array(1);/**
 @param {!TypedArray} input
 @param {string} original
 @param {string} target
 @param {!TypedArray} output
 */
function bitDepth(input,original,target,output){validateBitDepth_(original);validateBitDepth_(target);/** @type {!Function} */ var toFunction=getBitDepthFunction_(original,target);/** @type {!Object<string,number>} */ var options={oldMin:Math.pow(2,parseInt(original,10))/2,newMin:Math.pow(2,parseInt(target,10))/2,oldMax:Math.pow(2,parseInt(original,10))/2-1,newMax:Math.pow(2,parseInt(target,10))/2-1};/** @const @type {number} */ var len=input.length;if(original=="8")for(var i=0;i<len;i++)output[i]=
input[i]-=128;for(var i$0=0;i$0<len;i$0++)output[i$0]=toFunction(input[i$0],options);if(target=="8")for(var i$1=0;i$1<len;i$1++)output[i$1]=output[i$1]+=128}/**
 @private
 @param {number} sample
 @param {!Object<string,number>} args
 @return {number}
 */
function intToInt_(sample,args){if(sample>0)sample=parseInt(sample/args.oldMax*args.newMax,10);else sample=parseInt(sample/args.oldMin*args.newMin,10);return sample}/**
 @private
 @param {number} sample
 @param {!Object<string,number>} args
 @return {number}
 */
function floatToInt_(sample,args){return parseInt(sample>0?sample*args.newMax:sample*args.newMin,10)}/**
 @private
 @param {number} sample
 @param {!Object<string,number>} args
 @return {number}
 */
function intToFloat_(sample,args){return sample>0?sample/args.oldMax:sample/args.oldMin}/**
 @private
 @param {number} sample
 @return {number}
 */
function floatToFloat_(sample){f64f32_[0]=sample;return f64f32_[0]}/**
 @private
 @param {string} original
 @param {string} target
 @return {!Function}
 */
function getBitDepthFunction_(original,target){/** @type {!Function} */ var func=function(x){return x};if(original!=target)if(["32f","64"].includes(original))if(["32f","64"].includes(target))func=floatToFloat_;else func=floatToInt_;else if(["32f","64"].includes(target))func=intToFloat_;else func=intToInt_;return func}/**
 @private
 @param {string} bitDepth
 @throws {Error}
 */
function validateBitDepth_(bitDepth){if(bitDepth!="32f"&&bitDepth!="64"&&(parseInt(bitDepth,10)<"8"||parseInt(bitDepth,10)>"53"))throw new Error("Invalid bit depth.");}return bitDepth}();return bitDepth; })));
